File: internal/change/freeze_override.go
// SPDX-License-Identifier: Apache-2.0

// freeze_override.go implements T-4006's audited escape hatch for a
// declared freeze-window policy rule: a reasoned, recorded override that
// downgrades a freeze rule's otherwise-blocking `deny` finding to a
// visible warning — following the exact shape T-2604's two-person
// break-glass already established (twoperson.go): a written reason is
// required, the invocation is recorded server-side, and it is audited
// under its own action so a filtered audit log finds every override
// without knowing which validate/apply results imply one.
//
// WHY THIS IS NOT twoperson.go's break-glass, reused. Two-person
// break-glass overrides an AUTHORIZATION check (enforceTwoPerson, run once,
// only inside beginApply, only after validation has already passed). A
// freeze window is not an authorization check — it is an ordinary
// VALIDATE-time policy finding (policyValidate), by the task card's own
// requirement: a freeze must refuse a changeset "before an operator
// invests in staging and review, not at the moment of apply." So this
// override has to be visible to EVERY path that evaluates policy for a
// specific changeset — Service.validate (draft mutation, POST .../validate,
// and Schedule's pre-schedule validate), beginApply's pre-apply
// revalidation, and policyDenial's early Diff refusal — not only to
// beginApply. Reusing changeset_breakglass/change.breakglass for a
// conceptually different override would also make an auditor's "which
// ceremony happened here" question ambiguous (see the migration's own
// comment).
//
// ONE OVERRIDE RECORD, SEEN IDENTICALLY EVERYWHERE. overriddenPolicyTags is
// the single read every one of those paths calls (via
// Service.validationInputs / policyDenial) — so a freeze declared, then
// overridden, then edited-and-not-reinvoked can never look bypassed to one
// caller and blocked to another.

package change

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/bgovanlu/vnprox/internal/store"
)

// PolicyTagFreeze is the reserved policy-rule tag convention a freeze
// window declares itself with (`tags: [freeze]`). It is a convention, not
// a new field on PolicyRule or a new rule type: InvokeFreezeOverride
// downgrades a violation only when the violating rule carries this tag
// (via PolicyInput.OverriddenTags, policy_eval.go), and freeze_calendar.go's
// best-effort renderer uses the same tag to find freeze rules among an
// arbitrary policy set. A deployment is free to also use PolicyDeny/
// PolicyWarn rules with `time.*` conditions WITHOUT this tag (e.g. "warn
// outside business hours") — those are ordinary policy rules, just not
// ones the freeze override or the calendar recognize as a freeze window.
const PolicyTagFreeze = "freeze"

// maxFreezeOverrideReasonLen bounds the stored reason, mirroring
// maxBreakGlassReasonLen for the identical reason: a justification, not a
// blob store.
const maxFreezeOverrideReasonLen = 1000

// FreezeOverrideRecord is one changeset's freeze-window override, as the
// API renders it.
type FreezeOverrideRecord struct {
	ChangesetID string `json:"changesetId"`
	Reason      string `json:"reason"`
	InvokedBy   string `json:"invokedBy"`
	// OpsFingerprint pins the override to the ops it was invoked for.
	// Never rendered on the wire — an internal interlock, not information
	// an operator acts on (mirrors BreakGlassRecord's identical field).
	OpsFingerprint string `json:"-"`
	InvokedAt      int64  `json:"invokedAt"`
}

// ErrFreezeOverrideReasonRequired is returned by InvokeFreezeOverride when
// no written reason was supplied.
type ErrFreezeOverrideReasonRequired struct{}

func (e *ErrFreezeOverrideReasonRequired) Error() string {
	return "change: freeze-window override requires a written reason"
}

// ErrFreezeOverrideNotConfigured is returned by the freeze-override API
// when this Service was built with no freeze-override store wired.
type ErrFreezeOverrideNotConfigured struct{}

func (e *ErrFreezeOverrideNotConfigured) Error() string {
	return "change: freeze-override storage is not configured on this Service"
}

// InvokeFreezeOverride records changeset id's freeze-window override,
// attributed to actor and justified by reason (required —
// *ErrFreezeOverrideReasonRequired otherwise, before anything is written
// or audited).
//
// It is audited under its OWN action, `change.freeze_override` —
// deliberately not folded into `changeset.validate`/`changeset.apply`'s own
// audit entries, for the same "an auditor filtering for overrides finds
// them without knowing which result implies one" reason T-2604's
// break-glass gives.
//
// This does NOT itself validate or apply the changeset: the very next
// validate/apply/Diff call for id sees the override (via
// overriddenPolicyTags), and every OTHER policy rule — including any
// non-freeze deny — still applies in full. It downgrades exactly the
// tagged freeze rule(s)' findings from blocking to visible-and-warning.
func (s *Service) InvokeFreezeOverride(ctx context.Context, id, actor, reason string) (FreezeOverrideRecord, error) {
	if s.freezeOverrides == nil {
		return FreezeOverrideRecord{}, &ErrFreezeOverrideNotConfigured{}
	}
	reason = strings.TrimSpace(reason)
	if reason == "" {
		return FreezeOverrideRecord{}, &ErrFreezeOverrideReasonRequired{}
	}
	if len(reason) > maxFreezeOverrideReasonLen {
		return FreezeOverrideRecord{}, fmt.Errorf("change: freeze-override reason exceeds %d characters", maxFreezeOverrideReasonLen)
	}
	cs, err := s.Get(ctx, id)
	if err != nil {
		return FreezeOverrideRecord{}, err
	}
	now := s.now().Unix()
	row := store.ChangesetFreezeOverride{
		ChangesetID:    id,
		Reason:         reason,
		InvokedBy:      actor,
		InvokedAt:      now,
		OpsFingerprint: opsFingerprint(cs.Ops),
	}
	if err := s.freezeOverrides.Upsert(ctx, row); err != nil {
		return FreezeOverrideRecord{}, fmt.Errorf("change: recording freeze override for changeset %s: %w", id, err)
	}
	rec := freezeOverrideFromRow(row)
	s.appendAudit(ctx, actor, "change.freeze_override", "invoked", id, map[string]any{
		"reason": reason,
	})
	s.log.Warn("change: freeze-window override invoked",
		"changeset_id", id, "actor", actor, "reason", reason)
	return rec, nil
}

func freezeOverrideFromRow(row store.ChangesetFreezeOverride) FreezeOverrideRecord {
	return FreezeOverrideRecord{
		ChangesetID:    row.ChangesetID,
		Reason:         row.Reason,
		InvokedBy:      row.InvokedBy,
		InvokedAt:      row.InvokedAt,
		OpsFingerprint: row.OpsFingerprint,
	}
}

// freezeOverrideFor returns changesetID's override, or (zero, false) when
// none is on record. A read error is reported as "no override" AND logged
// — the consequence is a freeze that still blocks, never one that silently
// stops blocking (mirrors breakGlassFor's identical fail-closed contract).
func (s *Service) freezeOverrideFor(ctx context.Context, changesetID string) (FreezeOverrideRecord, bool) {
	if s.freezeOverrides == nil {
		return FreezeOverrideRecord{}, false
	}
	row, err := s.freezeOverrides.Get(ctx, changesetID)
	if err != nil {
		if !errors.Is(err, store.ErrNotFound) {
			s.log.Error("change: reading freeze-override record", "changeset_id", changesetID, "error", err)
		}
		return FreezeOverrideRecord{}, false
	}
	return freezeOverrideFromRow(row), true
}

// FreezeOverrideEvents returns every recorded freeze-window override,
// newest first — mirrors BreakGlassEvents' contract so a future findings
// check (T-4007's maintenance-window suppression is the next card to touch
// this calendar/window model) can compute visibility from these rows the
// same way change_break_glass already does from BreakGlassEvents, without
// this file needing to know anything about internal/findings.
func (s *Service) FreezeOverrideEvents(ctx context.Context) []FreezeOverrideRecord {
	if s.freezeOverrides == nil {
		return nil
	}
	rows, err := s.freezeOverrides.List(ctx)
	if err != nil {
		s.log.Warn("change: listing freeze-override records", "error", err)
		return nil
	}
	out := make([]FreezeOverrideRecord, 0, len(rows))
	for _, row := range rows {
		out = append(out, freezeOverrideFromRow(row))
	}
	return out
}

// overriddenPolicyTags builds the PolicyInput.OverriddenTags a changeset's
// policy evaluation needs: when changesetID has a freeze override on
// record AND it is still pinned to ops (an edit after the override
// invalidates it — the same rule twoperson.go's break-glass follows via
// its own OpsFingerprint check in enforceTwoPerson), every freeze-tagged
// rule's otherwise-blocking finding is downgraded, with a note naming who
// overrode it and why.
//
// changesetID == "" (a changeset that does not exist yet — Create/
// CreateRequest) always returns nil: there is nothing an override could
// have been recorded against.
func (s *Service) overriddenPolicyTags(ctx context.Context, changesetID string, ops []Op) map[string]string {
	if changesetID == "" {
		return nil
	}
	fo, ok := s.freezeOverrideFor(ctx, changesetID)
	if !ok || fo.OpsFingerprint != opsFingerprint(ops) {
		return nil
	}
	return map[string]string{
		PolicyTagFreeze: fmt.Sprintf("freeze override by %s: %s", fo.InvokedBy, fo.Reason),
	}
}

File: web/src/changesets/freezeOverride.ts
// SPDX-License-Identifier: Apache-2.0

// T-4006: `POST /changesets/{id}/freeze-override` — the audited escape
// hatch for a declared freeze-window policy rule, following the exact
// shape T-2604's break-glass established (breakGlass.ts): a written reason
// is required server-side, the override is recorded and audited under its
// own action, and it is pinned to the ops it was taken for.
//
// UNLIKE break-glass, this override does not gate an authorization check —
// a freeze window is an ordinary VALIDATE-time policy finding, so
// overriding it does not remove the finding. It downgrades it to a visible
// warning naming the override. The finding staying on screen, annotated,
// IS the "audited, not silent" property: an operator applying under an
// override still sees exactly what they overrode.
import { apiFetch } from "../api/client";
import { readCsrfCookie } from "../api/auth";
import type { Finding, FreezeOverrideRecord } from "../api/types";
import type { PolicyResult } from "../api/policies";

/** The reserved policy-rule tag convention a freeze window declares itself
 * with (internal/change.PolicyTagFreeze). */
export const FREEZE_TAG = "freeze";

/** Whether the panel should be offered: the changeset carries a currently-
 * blocking (`severity: error`) `policy.violation` finding — the ONE ground
 * truth for "is a deny rule actually stopping this apply right now", since
 * an already-overridden rule's finding is downgraded to `warning`
 * server-side (policy_eval.go's policyFinding) — AND `POST /policies/test`
 * (usePolicyVerdictQuery, already fetched for PolicyVerdictPanel) shows a
 * freeze-tagged rule among the violations.
 *
 * The two are checked independently rather than correlated 1:1 by rule id:
 * the wire `Finding` carries no structured `ruleId` (policyVerdict.ts's own
 * doc comment explains why this client never parses one out of the message
 * string), so this is a best-effort UI HINT, never an enforcement decision
 * — offering the override when the real cause is a different deny rule
 * costs nothing (the block simply persists, visibly, if it does not help);
 * never offering it when a freeze rule truly is blocking would be the
 * failure that matters, so this deliberately leans toward offering it. */
export function freezeBlocksApply(findings: readonly Finding[], result: PolicyResult | undefined): boolean {
  const hasBlockingPolicyFinding = findings.some((f) => f.code === "policy.violation" && f.severity === "error");
  if (!hasBlockingPolicyFinding) return false;
  return (result?.rules ?? []).some(
    (r) => (r.tags ?? []).includes(FREEZE_TAG) && (r.violatingOps ?? []).length > 0,
  );
}

/** The server's own bound on a stored reason
 * (internal/change.maxFreezeOverrideReasonLen). */
export const MAX_FREEZE_OVERRIDE_REASON_LEN = 1000;

/** What invoking it does, stated before the confirm control — mirrors
 * BREAK_GLASS_CONSEQUENCES's role exactly, adjusted for the one real
 * difference: this does not touch an authorization gate, it downgrades a
 * validate-time finding to a visible warning. */
export const FREEZE_OVERRIDE_CONSEQUENCES: readonly string[] = [
  "An audit entry is written under its own action, `change.freeze_override`, naming you, the changeset and the reason you type.",
  "The freeze rule's finding stays on the changeset — it is downgraded from blocking to a visible warning that names the override, never silently removed.",
  "It affects ONLY the freeze-tagged rule(s) that matched. Any other policy deny, validation error, or approval requirement still blocks exactly as before.",
  "It is pinned to the operations it was taken for. Editing the draft afterwards does not carry it over — the freeze blocks again, and a fresh override has to be taken.",
];

/** Whether a typed reason is one the server will accept. */
export function freezeOverrideReasonError(reason: string): string | undefined {
  const trimmed = reason.trim();
  if (trimmed.length === 0) {
    return "A written reason is required. An override with no justification is exactly what this ceremony exists to prevent.";
  }
  if (trimmed.length > MAX_FREEZE_OVERRIDE_REASON_LEN) {
    return `The reason must be at most ${String(MAX_FREEZE_OVERRIDE_REASON_LEN)} characters; this one is ${String(trimmed.length)}.`;
  }
  return undefined;
}

/** POST /changesets/{id}/freeze-override — `netWrite` + CSRF. Records the
 * override; the very next validate/apply/diff decides, server-side, whether
 * it applies to the ops actually staged. `503 freeze_override_unavailable`
 * on a daemon with no freeze-override store, a deployment fact rather than
 * a refusal of this particular changeset. */
export function invokeFreezeOverride(id: string, reason: string): Promise<FreezeOverrideRecord> {
  return apiFetch<FreezeOverrideRecord>(`/changesets/${encodeURIComponent(id)}/freeze-override`, {
    method: "POST",
    json: { reason: reason.trim() },
    csrfToken: readCsrfCookie(),
  });
}
